src/orange/orangeinp/PolySolid.cc

```diff
@@ -0,0 +1,252 @@
+{
+    if (auto const& sea = base.enclosed_angle())
+    {
+        // The enclosed angle is "true" (specified by the user to truncate the
+        // shape azimuthally): construct a wedge to be added or deleted
+        auto&& [sense, wedge] = sea.make_wedge();
+        NodeId wedge_id
+            = build_intersect_region(vb, base.label(), "angle", wedge);
+        if (sense == Sense::outside)
+        {
+            wedge_id = vb.insert_region({}, Negated{wedge_id});
+        }
+        result
+            = vb.insert_region(Label{std::string{base.label()}, "restricted"},
+                               Joined{op_and, {result, wedge_id}});
+    }
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from a filled polygon solid.
+ */
+PolySegments::PolySegments(VecReal&& outer, VecReal&& z)
+    : PolySegments{{}, std::move(outer), std::move(z)}
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from a shell of a polygon solid.
+ */
+PolySegments::PolySegments(VecReal&& inner, VecReal&& outer, VecReal&& z)
+    : inner_{std::move(inner)}, outer_{std::move(outer)}, z_{std::move(z)}
+{
+    CELER_VALIDATE(z_.size() >= 2,
+                   << "no axial segments was specified: at least 2 points "
+                      "needed (given "
+                   << z_.size() << ")");
+    CELER_VALIDATE(outer_.size() == z_.size(),
+                   << "inconsistent outer radius size (" << outer_.size()
+                   << "): expected " << z_.size());
+    CELER_VALIDATE(inner_.empty() || inner_.size() == z_.size(),
+                   << "inconsistent inner radius size (" << inner_.size()
+                   << "): expected " << z_.size());
+
+    CELER_VALIDATE(is_monotonic_nondecreasing(make_span(z_)),
+                   << "axial grid has increasing grid points");
+    for (auto i : range(outer_.size()))
+    {
+        CELER_VALIDATE(outer_[i] >= 0, << "invalid outer radius " << outer_[i]);
+        CELER_VALIDATE(inner_.empty()
+                           || (inner_[i] >= 0 && inner_[i] <= outer_[i]),
+                       << "invalid inner radius " << inner_[i]);
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build with label, axial segments, optional restriction.
+ */
+PolySolidBase::PolySolidBase(std::string&& label,
+                             PolySegments&& segments,
+                             SolidEnclosedAngle&& enclosed)
+    : label_{std::move(label)}
+    , segments_{std::move(segments)}
+    , enclosed_{std::move(enclosed)}
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return a polycone *or* a simplified version for only a single segment.
+ */
+auto PolyCone::or_solid(std::string&& label,
```

Comment:
Personally I think `solid_or_shape` would be better than `or_solid`, going all the way back up to `Solid<T>::or_shape`.